Select multiple points at person modal

paolodamico opened the pull request:
## Changes

Closes #4775. You can now pick a different data point to visualize in the person modal. Particularly helpful for when there are overlapping data points in your graph. This initial version only targets regular trends graphs (it is only where we have this problem, we may want to support elsewhere for consistency, but for a later improvement).

![image](https://user-images.githubusercontent.com/5864173/145467451-e69d50d4-0c47-4772-b939-4d0e64141a05.gif)

**With breakdowns**
<img width="601" alt="" src="https://user-images.githubusercontent.com/5864173/145467561-e7dbdb8c-9fb1-4ef8-b15c-5ac56b61cbba.png">


## How did you test this code?
TBD.

paolodamico committed:
send crossDataset to person modal

paolodamico committed:
basic label display

paolodamico committed:
selected id

paolodamico committed:
improve the UI

paolodamico committed:
switch data points

paolodamico committed:
ts adjustment

paolodamico committed:
small bf

paolodamico commented:
I think if we implement that we could just add an extra option on the dropdown to view all for that given date.

paolodamico commented:
Ping for a review today @alexkim205 if you can

alexkim205 commented on frontend/src/lib/components/InsightLabel/index.tsx:
```diff
@@ -133,23 +133,34 @@ export function InsightLabel({
+                                    ))}
+                                </Space>
+                            ) : (
+                                <>
```
I think @EDsCODE made a fix here already that needs to be resolved

alexkim205 commented on frontend/src/scenes/insights/InsightsTable/InsightsTable.tsx:
```diff
@@ -98,7 +98,7 @@ export function InsightsTable({
                 return (
                     <PHCheckbox
-                        color={colorList[item.id]}
+                        color={colorList[item.id % colorList.length]}
```
`getChartColors` also takes in a second `numSeries` where you can pass in the number of colors you want to generate here. I think it'll be better to pass in `items.length` there and have the util function handle any modulo wrapping

alexkim205 commented on frontend/src/scenes/insights/LineGraph/LineGraph.jsx:
```diff
@@ -402,6 +402,16 @@ export function LineGraph({
             onClick: (_, [point]) => {
                 if (point && onClick) {
                     const dataset = datasets[point._datasetIndex]
+                    const crossDataset = datasets
+                        .filter((_dt) => !_dt.dotted)
+                        .map((_dt) => ({
+                            seriesId: _dt.id,
+                            action: _dt.action,
+                            personUrl:
+                                typeof point._index !== 'undefined' ? _dt.persons_urls?.[point._index].url : undefined,
+                            breakdown_value: _dt.breakdown_value,
+                            pointValue: typeof point._index !== 'undefined' ? _dt.data[point._index] : undefined,
+                        }))
```
The new refactor of `LineGraph` has a more robust way of capturing which datasets are being clicked on. The `onClick` callback will now have `pointsIntersectingLine` and `pointsIntersectingClick` which are more robustly calculated by `chart.js's` `Chart.getElementsAtEventForMode()` function. I think you can leverage that to achieve the same effect here.

alexkim205 commented on frontend/src/scenes/trends/PersonsModal.tsx:
```diff
@@ -69,7 +77,7 @@ export function PersonsModal({
             ) : (
                 <>
-                    <PropertyKeyInfo value={people?.label || ''} disablePopover /> on{' '}
+                    {capitalizeFirstLetter(actorLabel)} list on{' '}
```
Nice, `PropertyKeyInfo` is getting pretty overloaded anyways

alexkim205 commented on frontend/src/scenes/trends/PersonsModal.tsx:
```diff
@@ -153,6 +165,34 @@ export function PersonsModal({
+                                                key={`${dataPoint.action.id}${dataPoint.breakdown_value}`}
+                                            >
+                                                <InsightLabel
+                                                    seriesColor={colorList[dataPoint.seriesId % colorList.length]}
```
The same can be said here about passing in `numSeries` into `getChartColors`

alexkim205 commented on frontend/src/scenes/trends/personsModalLogic.tsx:
```diff
@@ -418,8 +453,31 @@ export const personsModalLogic = kea<personsModalLogicType<LoadPeopleFromUrlProp
+        switchToDataPoint: async ({ seriesId }) => {
+            const data = values.people?.crossDataset?.find(({ seriesId: _id }) => _id === seriesId)
+
+            if (data && values.peopleParams) {
```
If these conditions aren't met, we run the risk of showing stale persons in the modal once the data point has been switched. This can be confusing since users might believe they are looking at the new data point they just switched to. We should either return an empty list of people or prevent these data points from being shown in the dropdown in the first place.

paolodamico replied:
I believe these conditions should always be met, as we only show elements in the dropdown that come from `crossDataset`, so `data` should always be defined in theory.

alexkim205 commented on frontend/src/scenes/trends/viz/ActionsPie.tsx:
```diff
@@ -87,6 +87,7 @@ export function ActionsPie({
                                           date_to,
                                           filters: filtersParam,
                                           breakdown_value,
+                                          seriesId,
```
I might not be reading this right, but isn't `seriesId` always the same as `point.index` in the context of pie charts?

alexkim205 reviewed (changes requested):
Hey @paolodamico, this looks like a nice feature to get in. However there're a few merge conflicts that came up as a result of the `LineGraph.jsx` refactor that should be addressed and some in-line feedback for your consideration.

On the plus side, I think it'd be okay to implement this only for the typed `LineGraph.tsx` version instead of making changes to both the new and old (js) files, since we'll be deprecating `LEGACY_LineGraph.jsx` soon.

paolodamico committed:
small fix

paolodamico committed:
Merge branch 'master' into person-modal-multiple

paolodamico committed:
try to fix everything

paolodamico committed:
fix everything

paolodamico committed:
Update PersonsModal.tsx

paolodamico committed:
simplify typing & logic

paolodamico commented:
Alright addressed all feedback & finally managed to get this working with the new `LineGraph.tsx` and all related changes. I've put the new feature behind a feature flag, and it does require to be running the new `LineGraph.tsx` component too, will merge to avoid any more merge conflicts, but feel free to open a new PR with improvements/suggestions. Thanks @alexkim205!

paolodamico committed:
remove stale console.log
